Shift modifiers v3

Screwtapello opened the pull request:
This branch teaches Kakoune that Shift is a possible modifier, including the `s-` prefix in key names. `<s-x>` in normalized to `<X>` when `x` is a lowercase ASCII character, but other printable characters (including aliases for printable characters like `lt`, `gt`, `plus` and `minus`) are disallowed because the behaviour of the shift modifier depends on the user's keyboard layout, which Kakoune cannot determine from inside a terminal.

Shifted keys that ncurses recognises are mapped to properly-modified keys so they can be mapped by users, and added to the default normal mappings where appropriate.

This still has some potential problems, but I don't think addressing them is practical:

  - It's possible to create a mapping for a key like `<c-s-a>`, but it cannot be triggered because terminals ignore the shift modifier for control keys
  - You can create a mapping for `<s-f11>`, but it cannot be triggered because there's no universal standard for what key-code ncurses will use for modified function keys
  - It is not universally true that applying the shift modifier to a lower-case ASCII letter will produce the corresponding uppercase ASCII letter (see [Dotted and dotless I](https://en.wikipedia.org/wiki/Dotted_and_dotless_I)), but Kakoune's default keymap already assumes this is the case for mnemonic reasons (`i` is insert, `I` is insert at the beginning of the line) so that ship has already sailed

This branch also includes a commit that teaches Kakoune to normalise `<s-tab>` to `<backtab>` in the same way that it normalises `<s-x>` to `<X>`. Backtab is [theoretically distinct](https://deskthority.net/wiki/Tab_key) from the tab key, but I bet it will be `<s-tab>` on every computer Kakoune is ever run on, and somebody wanting to bind it will probably guess `<s-tab>` before they guess `<backtab>`. However, I can imagine reasons not to do this, or not to do it this way, so I left it as a separate commit that can be left out if desired.

lenormf commented on src/keys.cc:
```diff
@@ -192,6 +231,35 @@ UnitTest test_keys{[]()
+    try {
+        parse_keys("<s-X>");
+    } catch (runtime_error& err) {
+	kak_assert(err.what() == "Shift modifier only works on special keys and lowercase ASCII");
```
Relying on error messages is error prone, wouldn't it be enough to check that an exception is thrown?

lenormf replied:
I think it would make sense to have a custom exception for parsing errors, @mawww ?

mawww replied:
yeah, introducing a specific exception type, inheriting from runtime_error for that is fine. But frankly, parse_keys has not much other ways to fail than failing to parse a key... Also, introduce a macro for that in unit_test.hh `kak_expect_throw(exception_type, expression)` to keep the unit test easy to follow.

mawww commented on src/keys.cc:
```diff
@@ -102,9 +103,41 @@ KeyList parse_keys(StringView str)
+        {
+            Codepoint key = name_it->key;
+
+            if (modifier & Key::Modifiers::Shift)
```
move that logic in canonicalize_ifn

mawww commented on src/keys.cc:
```diff
@@ -102,9 +103,41 @@ KeyList parse_keys(StringView str)
+            {
+                if (key < 0xD800 || key > 0xDFFF)
+                {
+                    if (!is_lower(name_it->key))
```
we can probably tolerate upper case as well `<s-X> == <s-x> == <X>`, use `is_basic_alpha` to check that.

mawww commented on src/keys.cc:
```diff
@@ -167,9 +200,13 @@ String key_to_str(Key key)
-    case Key::Modifiers::Control:    res = "c-" + res; named = true; break;
-    case Key::Modifiers::Alt:        res = "a-" + res; named = true; break;
-    case Key::Modifiers::ControlAlt: res = "c-a-" + res; named = true; break;
+    case Key::Modifiers::Control:         res = "c-" + res; named = true; break;
```
now that we have all those combination, a loop on Control, Alt, and Shift that prepend respectively "c-", "a-", and "s-" would be cleaner.

mawww commented on src/keys.hh:
```diff
@@ -19,17 +19,21 @@ struct Key
-        MousePos     = 1 << 4,
-        MouseWheelDown = 1 << 5,
-        MouseWheelUp = 1 << 6,
+        ControlShift = Control | Shift,
```
No need to add all those combined alternatives, I'd rather remove ControlAlt.

mawww commented on src/keys.hh:
```diff
@@ -109,6 +117,18 @@ constexpr Key ctrlalt(Key key)
 {
     return { key.modifiers | Key::Modifiers::ControlAlt, key.key };
 }
+constexpr Key ctrlshift(Key key)
```
same thing as for the modifiers, just let the user write `ctrl(alt(key))`, and remove `ctrlalt`

mawww commented on src/ncurses_ui.cc:
```diff
@@ -569,14 +569,23 @@ Optional<Key> NCursesUI::get_next_key()
         {
         case KEY_BACKSPACE: case 127: return {Key::Backspace};
         case KEY_DC: return {Key::Delete};
+        case KEY_SDC: return {{Key::Modifiers::Shift, Key::Delete}};
```
use `shift(Key::Delete)`

mawww commented on src/normal.cc:
```diff
@@ -2055,6 +2055,11 @@ static const HashMap<Key, NormalCmd, MemoryDomain::Undefined, KeymapBackend> key
     { {'K'}, {"extend up", move<LineCount, Backward, SelectMode::Extend>} },
     { {'L'}, {"extend right", move<CharCount, Forward, SelectMode::Extend>} },
 
+    { {Key::Modifiers::Shift, Key::Left}, {"extend left", move<CharCount, Backward, SelectMode::Extend>} },
```
`shift(Key::Left)`

Screwtapello commented:
I've pushed a new version of my changes, rebased on top of current master and squashed to a single commit as requested on IRC. I believe this addresses all the code-review comments so far, although I may have gone a little overboard on key-parsing exceptions: of the seven exception-types now in `exception.hh`, over half are specific ways in which key-parsing can fail. At least it makes the tests reliable!

mawww commented on src/exception.hh:
```diff
@@ -23,6 +23,30 @@ private:
     String m_what;
 };
 
+struct key_parse_error : runtime_error
```
This does not belong to exception.hh, which does not contain any specific option type. it should be directly in keys.cc as nothing else is going to know about it except the unit test (and I am still on the fence that we really need to introduce an exception for that, so please keep it to a single one)

mawww commented on src/exception.hh:
```diff
@@ -23,6 +23,30 @@ private:
 
+struct key_parse_error : runtime_error
+{
+    key_parse_error(String what)
```
`using runtime_error::runtime_error;` to inherit that constructor.

mawww commented on src/input_handler.cc:
```diff
@@ -1568,8 +1568,10 @@ InputHandler::ScopedForceNormal::~ScopedForceNormal()
 
 static bool is_valid(Key key)
 {
+    Key::Modifiers valid_mods = (Key::Modifiers::Control | Key::Modifiers::Alt | Key::Modifiers::Shift);
```
constexpr

mawww commented on src/keys.cc:
```diff
@@ -19,6 +19,28 @@ static Key canonicalize_ifn(Key key)
+            key.modifiers &= ~Key::Modifiers::Shift;
+            key.key = to_upper(key.key);
+        }
+        else if (key.key == Key::Tab)
```
You know what, this is a good occasion to get rid of backtab, and move the backtab == shift-tab hack into ncurses.

mawww commented on src/keys.cc:
```diff
@@ -165,13 +188,10 @@ String key_to_str(Key key)
-    case Key::Modifiers::ControlAlt: res = "c-a-" + res; named = true; break;
-    default: break;
-    }
+    if (key.modifiers & Key::Modifiers::Shift)   { res = "s-" + res; named = true; }
```
might be nitpicking, but I'd put them in the control, alt, shift order, as I tend to say "ctrl alt blah", "ctrl shift blah" "alt shift blah" rather than "alt ctrl blah" "shift ctrl blah" or "shift alt blah".

mawww replied:
You are right, I was confused.

mawww commented on src/ncurses_ui.cc:
```diff
@@ -570,14 +570,23 @@ Optional<Key> NCursesUI::get_next_key()
         {
         case KEY_BACKSPACE: case 127: return {Key::Backspace};
         case KEY_DC: return {Key::Delete};
+        case KEY_SDC: return {shift({Key::Delete})};
```
no need to braces here, `shift` already returns a Key.

mawww commented on src/normal.cc:
```diff
@@ -2062,6 +2062,11 @@ static const HashMap<Key, NormalCmd, MemoryDomain::Undefined, KeymapBackend> key
     { {'K'}, {"extend up", move<LineCount, Backward, SelectMode::Extend>} },
     { {'L'}, {"extend right", move<CharCount, Forward, SelectMode::Extend>} },
 
+    { {shift(Key::Left)}, {"extend left", move<CharCount, Backward, SelectMode::Extend>} },
```
same, no need for braces here

Screwtapello commented:
Once again, thanks for being patient with my C++ inexperience.

I've pushed an updated branch that addresses comments from the latest review, rebased and squashed again.

As suggested, I've removed all support for `<backtab>` and left `<s-tab>` as the One True Syntax for that particular key combination. For the sake of users who might have `<backtab>` mappings configured, I added a section to `mappings.asciidoc` to help people migrate. I started by normalizing `<backtab>` to `<s-tab>` while keeping support for the old name, but it started getting complicated so I just went with removal instead.

mawww commented on doc/pages/mapping.asciidoc:
```diff
@@ -110,3 +112,18 @@ be used:
+the way terminals handle control characters, mappings like *<c-s-a>* are
+unlikely to work in Kakoune's terminal UI.
+
+== Unsupported keys
```
The place to document breakage is the startup message in main. I dont want to litter the documentation with references to how things used to be. (Just a single sentence, "<backtab> is gone, use <s-tab>").

mawww commented on src/keys.cc:
```diff
@@ -182,16 +200,39 @@ UnitTest test_keys{[]()
+         { Key::Up },
          { Key::Modifiers::Alt, 'j' },
-         { Key::Modifiers::Control, 'r' }
+         { Key::Modifiers::Control, 'r' },
```
can you convert those to `alt('j'), ctrl('r'), and `shift(Key::Up)` while your are at it ?

mawww commented on src/ncurses_ui.cc:
```diff
@@ -570,15 +570,24 @@ Optional<Key> NCursesUI::get_next_key()
         {
         case KEY_BACKSPACE: case 127: return {Key::Backspace};
         case KEY_DC: return {Key::Delete};
+        case KEY_SDC: return shift({Key::Delete});
```
I dont think you need the braces inside the shift call.
